lib/galaxy/tools/parameters/grouping.py

```diff
@@ -219,17 +219,30 @@ def get_file_base_name(self, context):
         fd = context.get('files_metadata|base_name', 'Galaxy_Composite_file')
         return fd
 
-    def get_file_type(self, context):
-        return context.get(self.file_type_name, self.default_file_type)
-
-    def get_datatype_ext(self, trans, context):
-        ext = self.get_file_type(context)
+    def get_file_type(self, context, parent_context=None):
+        file_type = context.get(self.file_type_name, None)
+        if file_type == "":
+            if parent_context:
+                file_type = parent_context.get(self.file_type_name, self.default_file_type)
+            else:
+                file_type = self.default_file_type
+        return file_type
+
+    def get_dbkey(self, context, parent_context=None):
+        dbkey = context.get("dbkey", None)
+        if dbkey == "":
+            if parent_context:
+                dbkey = parent_context.get("dbkey", dbkey)
+        return dbkey
```

Comment:
I might be missing something but why are we not just returning `parent_context.get("dbkey", context.get("dbkey"))`?

Reply:
So the inner dbkey - the per file one - is just a ``hidden`` field in the tool form with a default of ``""`` - that is what the outer one is checking the top-level ``dbkey`` param in the tool that I think will default to ``?``. So if the API request sends a ``dbkey`` for all files and then one for a specific file - I think we should use the base for all files that don't specify an explicit ``dbkey`` and then use the specific keys as set. I tried to use this strategy with the file extensions also. This implementation is crap because we are mapping API requests to a tool that wasn't really ever designed to do this well (these frustrations made led me to create #4734).

So I implemented API tests to verify all this behavior with respect to specific file dbkey versus dbkey for all files - if you can rework the implementation to be cleaner in such a way that the API calls don't break - I'm totally on board. I'm quite frustrated with a lot of ``grouping.py``. There are more test cases in #4746 that should also help verify cleanups to the implementation don't break FTP uploads and such.